affinity/cpp-20/d0796r2.md

```diff
@@ -110,25 +113,39 @@ There are also some additional challenges which we have been investigating but a
+The first task in allowing C++ applications to leverage memory locality is to provide the ability to query a *system* for its *resource topology* (commonly represented as a tree or graph) and traverse its *execution resources* and *memory resources*.
 
 The capability of querying underlying *execution resources* of a given *system* is particularly important towards supporting affinity control in C++. The current proposal for executors [[22]][p0443r4] leaves the *execution resource* largely unspecified. This is intentional: *execution resources* will vary greatly between one implementation and another, and it is out of the scope of the current executors proposal to define those. There is current work [[23]][p0737r0] on extending the executors proposal to describe a typical interface for an *execution context*. In this paper a typical *execution context* is defined with an interface for construction and comparison, and for retrieving an *executor*, waiting on submitted work to complete and querying the underlying *execution resource*. Extending the executors interface to provide topology information can serve as a basis for providing a unified interface to expose affinity. This interface cannot mandate a specific architectural definition, and must be generic enough that future architectural evolutions can still be expressed.
 
-Two important considerations when defining a unified interface for querying the *resource topology* of a *system*, are (a) what level of abstraction such an interface should have, and (b) at what granularity it should describe the topology's *execution resources*. As both the level of abstraction of an *execution resource* and the granularity that it is described in will vary greatly from one implementation to another, it’s important for the interface to be generic enough to support any level of abstraction. To achieve this we propose a generic hierarchical structure of *execution resources*, each *execution resource* being composed of other *execution resources* recursively. Each *execution resource* within this hierarchy can be used to place memory (i.e., allocate memory within the *execution resource’s* memory region), place execution (i.e. bind an execution to an *execution resource’s execution agents*), or both.
+Two important considerations when defining a unified interface for querying the *resource topology* of a *system*, are (a) what level of abstraction such an interface should have, and (b) at what granularity it should describe the topology's *execution resources* and *memory resources*. As both the level of abstraction of resources and the granularity that they describe in will vary greatly from one implementation to another, it’s important for the interface to be generic enough to support any level of abstraction. To achieve this we propose generic hierarchical structures for *execution resources* and *memory resources*, where each *resource* being composed of other *resources* recursively. 
```

Comment:
I think we hould state here that each execution resource should be made up of other execution resources and each memory resource should be made up of other memory resources:

"where each *resource* being composed of other *resources* recursively" -> "where each is recursively composed of other *execution resources*  and *memory resources* respectively"

Reply:
I started down this path and the problem I ran into was having purely hierarchical memory resource.  Even without GPU in the picture you will have main memory, high bandwidth memory, and non-volatile memory.  A hierarchical memory resource would require a top-level allocator that spans all of them.  The path I followed was instead for each execution resource to have a set of memory resources.  
I am uneasy rushing this scope, a little more design-discussion time would have been helpful but we are up against the pre-meeting mailing clock.

Reply:
That's a good point, having a separate set of memory resources for each execution resource could work, though this could lead to many execution resources pointing to the same memory resource, which could be confusing for users to figure out which memory resources are the same and which ones are different. Though perhaps we could maintain a hierarchy of memory resources, but have each execution resource point to the highest level memory resource which it can allocate memory in, and then the lower level memory resources could be traversed from that it would be consistent enough as to avoid confusion.

One way we could get around the problem of needing to have a top-level allocator could be to have a memory resource which represents the top of the topology which is either not capable of allocating (though that would likely mean throwing an exception) or is free to allocate anywhere within the topology.

I agree I think the separation of execution and memory topologies has revealed some difficult design points which I think we will need some further discussion to resolve.